polkadot/node/core/provisioner/src/lib.rs

```diff
@@ -643,60 +652,119 @@ async fn request_backable_candidates(
+					bitfields,
+					&occupied_core.availability,
+				);
+
+				if is_available {
+					ancestors
+						.entry(occupied_core.para_id())
+						.or_default()
+						.entry(occupied_core.candidate_hash)
+						.or_default()
+						.record();
+
 					if let Some(ref scheduled_core) = occupied_core.next_up_on_available {
-						// The candidate occupying the core is available, choose its
-						// child in the fragment tree.
-						//
-						// TODO: doesn't work for on-demand parachains. We lean hard on the
-						// assumption that cores are fixed to specific parachains within a session.
-						// https://github.com/paritytech/polkadot/issues/5492
-						(scheduled_core.para_id, vec![occupied_core.candidate_hash])
-					} else {
-						continue
+						// Request a new backable candidate for the newly scheduled para id.
+						scheduled_cores
+							.entry(scheduled_core.para_id)
+							.or_insert(HashSet::new())
+							.insert(core_idx);
 					}
+				} else if occupied_core.time_out_at > block_number {
+					// Not timed out and not available.
+					ancestors
+						.entry(occupied_core.para_id())
+						.or_default()
+						.entry(occupied_core.candidate_hash)
+						.or_default()
+						.record();
 				} else {
-					if occupied_core.time_out_at != block_number {
-						continue
-					}
+					// Timed out before being available.
+					let ancestor_state = ancestors
+						.entry(occupied_core.para_id())
+						.or_default()
+						.entry(occupied_core.candidate_hash)
+						.or_default();
+					ancestor_state.record();
+					ancestor_state.timeout();
+
 					if let Some(ref scheduled_core) = occupied_core.next_up_on_time_out {
 						// Candidate's availability timed out, practically same as scheduled.
-						(scheduled_core.para_id, Vec::new())
-					} else {
-						continue
+						scheduled_cores
+							.entry(scheduled_core.para_id)
+							.or_insert(HashSet::new())
+							.insert(core_idx);
 					}
 				}
 			},
 			CoreState::Free => continue,
 		};
+	}
 
-		let response = get_backable_candidate(relay_parent, para_id, required_path, sender).await?;
+	'para_loop: for (para_id, cores) in scheduled_cores {
+		let para_ancestors = ancestors.remove(&para_id).unwrap_or_default();
+		let core_count = cores.len();
+		let response = get_backable_candidates(
+			relay_parent,
+			para_id,
+			para_ancestors,
+			core_count as u32,
+			sender,
+		)
+		.await?;
 
-		match response {
-			Some((hash, relay_parent)) => selected_candidates.push((hash, relay_parent)),
-			None => {
-				gum::debug!(
-					target: LOG_TARGET,
-					leaf_hash = ?relay_parent,
-					core = core_idx,
-					"No backable candidate returned by prospective parachains",
-				);
-			},
+		if response.is_empty() {
+			gum::debug!(
+				target: LOG_TARGET,
+				leaf_hash = ?relay_parent,
+				?para_id,
+				"No backable candidate returned by prospective parachains",
+			);
+			continue
+		}
+
+		for (core_index, candidate) in cores.into_iter().zip(response.into_iter()) {
```

Comment:
I am only using the core mapping here because the runtime expects the vector of BackedCandidates to be sorted by CoreIndex, that's it. So we're not breaking anything with these changes.

As discussed, we may want to have this sorted in dependency order instead. But this will come in the future, when making runtime changes.

I agree with your larger point that it makes sense to decouple the backing validator group from the availability core index in the future

Reply:
> I am only using the core mapping here because the runtime expects the vector of BackedCandidates to be sorted by CoreIndex, that's it. So we're not breaking anything with these changes. 

but it is not matching the actual core right? We are just picking candidates from prospective parachains that match the para, we don't even know on which core they have been provided.

Reply:
Correct

Reply:
I removed this. We actually don't care whether or not the backed candidates are sorted in order of core_index, because they're sorted in the runtime: https://github.com/paritytech/polkadot-sdk/blob/ce5de99540eb1ceaed4b0a5d085983c4a273fdda/polkadot/runtime/parachains/src/paras_inherent/mod.rs#L1023